tpm docs (#3071)
hashicorp/web-unified-docs: content/vault/v2.x/content/api-docs/secret/identity/tpm-group.mdx

      "region": "us-east-1"
    },
    "name": "datacenter-east-tpms"
  }
}
```

## Update TPM group by ID

Update an existing TPM group by its identifier.

Providing `member_tpm_ids` replaces the existing member list entirely. Providing
`metadata` replaces the existing metadata map entirely.

| Method | Path                         |
| :----- | :--------------------------- |
| `POST` | `/identity/tpmgroup/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM group.

- `name` `(string: <optional>)` – Updated name for the group. Must be unique
  within the namespace.

- `metadata` `(key-value-map: {})` – Updated metadata. Replaces existing
  metadata entirely.

- `member_tpm_ids` `(comma-separated string list: [])` – Updated list of
  member TPM IDs. Replaces the existing member list entirely. Every ID must
  reference an existing TPM record in the current namespace.

### Sample payload

```json
{
  "metadata": {
    "environment": "staging"
  },
  "member_tpm_ids": [
    "sha256-a3f1c2d4e5b6..."
  ]
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
    --request POST \
    --data @payload.json \

>     ${VAULT_ADDR}/v1/identity/tpmgroup/id/f47ac10b-58cc-4372-a567-0e02b2c3d479

Author: What about this UUID?  Should it be all zeroes too?

Reviewer: Yes, please :)

---
layout: api
page_title: 'Identity Secret Backend: TPM Group - HTTP API'

> description: This is the API documentation for managing TPM groups in the identity store.

Reviewer: ```suggestion
description: API documentation for managing TPM groups in the identity store.
```

# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2026-07-02T00:00:00.000Z
last_modified: 2026-07-02T00:00:00.000Z
# END AUTO GENERATED METADATA
---

<EnterpriseAlert />

# Identity secrets engine - TPM group (API)

A TPM group organizes [TPM records](/vault/api-docs/secret/identity/tpm) into
logical collections for authorization purposes. Groups have a UUID-based ID,
a human-friendly name, optional metadata, and a list of member TPM IDs.

Key characteristics of TPM groups:

- The **ID** is a randomly generated UUID (unlike TPM records, whose ID is
  derived from the EK public key).
- If a `name` is not provided on creation, Vault generates a default name in
  the format `tpmgroup_<random-hex>`.
- TPM group names must be unique within a namespace. Attempting to create a
second group with a duplicate name returns an error.
- All TPM IDs listed in `member_tpm_ids` must already exist in the same

>   namespace; supplying an unknown ID returns an error.

Reviewer: ```suggestion
  namespace. Requests for an unknown ID return an error.
```

- Supplying `member_tpm_ids` on an update **replaces** the full member list
  rather than merging.
- Supplying `metadata` on an update **replaces** the full metadata map rather
  than merging.

## Create a TPM group

This endpoint creates a new TPM group. All `member_tpm_ids` values must

> reference existing TPM records in the same namespace.

Reviewer: ```suggestion
Create a new TPM group from existing TPM IDs in the same namespace.
```

| Method | Path                  |
| :----- | :-------------------- |
| `POST` | `/identity/tpmgroup`  |

### Parameters

- `name` `(string: "tpmgroup-<random-hex>")` – A human-friendly name for the
  TPM group. Must be unique within the namespace.

- `id` `(string: <optional>)` – ID of an existing TPM group to update. If set,
  the request updates that group rather than creating a new one.

- `metadata` `(key-value-map: {})` – Metadata to be associated with the group.
  In the CLI, this parameter can be repeated multiple times and all values are

>   merged together.

Reviewer: ```suggestion
  In CLI commands, you can repeat the `metadata` parameter multiple times and Vault will merge them together to create a single update string.
```
Can you not repeat the field in an API request body and have the same thing happen?

Author: In an API request body, the metadata map is represented as a map.  In the CLI, you have to say: `-metadata key1=val1 -metadata key2=val2`

Reviewer: Okay. I just wanted to confirm that folks couldn't provide multiple maps for metadata that we'd combine on their behalf. I don't know why anyone would want to construct an API call like that, but if the CLI allows it, I can see folks trying it 😅

- `member_tpm_ids` `(comma-separated string list: [])` – List of TPM record

>   IDs to assign as members of this group. Every ID must reference an existing

Reviewer: ```suggestion
  IDs to assign as members of the group. Every ID must reference an existing
```

  TPM record in the current namespace.

### Sample payload

```json
{
  "name": "datacenter-east-tpms",
  "metadata": {
    "environment": "production",
    "region": "us-east-1"
  },
  "member_tpm_ids": [
    "sha256-a3f1c2d4e5b6...",
    "sha256-7d8e9f0a1b2c..."
  ]
}
```

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     http://127.0.0.1:8200/v1/identity/tpmgroup

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup
```

```

### Sample response

```json
{
  "data": {
    "id": "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    "name": "datacenter-east-tpms"
  }
}
```

## Read TPM group by ID

> This endpoint queries a TPM group by its identifier.

Reviewer: ```suggestion
Query a TPM group by its identifier.
```

| Method | Path                         |
| :----- | :--------------------------- |
| `GET`  | `/identity/tpmgroup/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM group.

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

>     http://127.0.0.1:8200/v1/identity/tpmgroup/id/f47ac10b-58cc-4372-a567-0e02b2c3d479

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/id/00000000-0000-0000-0000-000000000000
```

```

### Sample response

```json
{
  "data": {

>     "id": "f47ac10b-58cc-4372-a567-0e02b2c3d479",

Reviewer: ```suggestion
    "id": "00000000-0000-0000-0000-000000000000",
```

    "member_tpm_ids": [
      "sha256-a3f1c2d4e5b6...",
      "sha256-7d8e9f0a1b2c..."
    ],
    "metadata": {
      "environment": "production",
      "region": "us-east-1"
    },
    "name": "datacenter-east-tpms"
  }
}
```

## Update TPM group by ID

> This endpoint updates an existing TPM group by its identifier.

Reviewer: ```suggestion
Update an existing TPM group by its identifier.
```

Providing `member_tpm_ids` replaces the existing member list entirely. Providing

> `metadata` replaces the existing metadata map entirely.

Reviewer: ```suggestion
Vault entirely replaces the existing member list with the value of `member_tpm_ids` and the group metadata map with the value of `metadata`.
```

Author: Your replacement text is a little ambiguous.  The original text was clear to me: if you provide either field, it will replace the current value.  In the new formulation, it could mean that, but it could also mean: if you fail to provide an explicit value for one of these fields, the existing value will be overridden with the default empty value for that field.

| Method | Path                         |
| :----- | :--------------------------- |
| `POST` | `/identity/tpmgroup/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM group.

- `name` `(string: <optional>)` – Updated name for the group. Must be unique
  within the namespace.

- `metadata` `(key-value-map: {})` – Updated metadata. Replaces existing
  metadata entirely.

- `member_tpm_ids` `(comma-separated string list: [])` – Updated list of
  member TPM IDs. Replaces the existing member list entirely. Every ID must
  reference an existing TPM record in the current namespace.

### Sample payload

```json
{
  "metadata": {
    "environment": "staging"
  },
  "member_tpm_ids": [
    "sha256-a3f1c2d4e5b6..."
  ]
}
```

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     http://127.0.0.1:8200/v1/identity/tpmgroup/id/f47ac10b-58cc-4372-a567-0e02b2c3d479

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/id/f47ac10b-58cc-4372-a567-0e02b2c3d479
```

```

### Sample response

```json
{
  "data": {

>     "id": "f47ac10b-58cc-4372-a567-0e02b2c3d479",

Reviewer: ```suggestion
    "id": "00000000-0000-0000-0000-000000000000",
```

    "name": "datacenter-east-tpms"
  }
}
```

## Delete TPM group by ID

This endpoint deletes a TPM group by its identifier. Deleting a group that does

> not exist is idempotent and returns no error.

Reviewer: ```suggestion
Deletes a TPM group by its identifier. Deleting a group that does
not exist is idempotent and returns successfully as a no-op.
```

| Method   | Path                         |
| :------- | :--------------------------- |
| `DELETE` | `/identity/tpmgroup/id/:id`  |

### Parameters

- `id` `(string: <required>)` – Identifier of the TPM group.

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request DELETE \

>     http://127.0.0.1:8200/v1/identity/tpmgroup/id/f47ac10b-58cc-4372-a567-0e02b2c3d479

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/id/f47ac10b-58cc-4372-a567-0e02b2c3d479
```

```

## List TPM groups by ID

This endpoint returns a list of all TPM group identifiers in the current

> namespace.

Reviewer: ```suggestion
Fetch a list of all TPM group identifiers in the current namespace.
```

| Method | Path                               |
| :----- | :--------------------------------- |
| `LIST` | `/identity/tpmgroup/id`            |
| `GET`  | `/identity/tpmgroup/id?list=true`  |

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request LIST \

>     http://127.0.0.1:8200/v1/identity/tpmgroup/id

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/id
```

```

### Sample response

```json
{
  "data": {
    "keys": [

>       "f47ac10b-58cc-4372-a567-0e02b2c3d479",

Reviewer: ```suggestion
      "00000000-0000-0000-0000-000000000000",
```

Author: This is just a UUID, is that something that sets off security scanner too?

Author: I re-read your original comment from https://github.com/hashicorp/web-unified-docs/pull/3071#discussion_r3754066842, and I see you did mean UUIDs too.  Not a fan but oh well, I'll change them.

>       "3c4d5e6f-7a8b-9c0d-1e2f-3a4b5c6d7e8f"

Reviewer: ```suggestion
      "11111111-1111-1111-1111-111111111111"
```

    ]
  }
}
```

## Read TPM group by name

> This endpoint queries a TPM group by its name.

Reviewer: ```suggestion
Fetch TPM group details by name.
```

| Method | Path                             |
| :----- | :------------------------------- |
| `GET`  | `/identity/tpmgroup/name/:name`  |

### Parameters

- `name` `(string: <required>)` – Name of the TPM group.

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

>     http://127.0.0.1:8200/v1/identity/tpmgroup/name/datacenter-east-tpms

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/name/datacenter-east-tpms
```

```

### Sample response

```json
{
  "data": {

>     "id": "f47ac10b-58cc-4372-a567-0e02b2c3d479",

Reviewer: ```suggestion
    "id": "00000000-0000-0000-0000-000000000000",
```

    "member_tpm_ids": [
      "sha256-a3f1c2d4e5b6..."
    ],
    "metadata": {
      "environment": "staging"
    },
    "name": "datacenter-east-tpms"
  }
}
```

## Update TPM group by name

This endpoint updates an existing TPM group by its name. The group must already

> exist; this endpoint does not create new groups.

Reviewer: ```suggestion
Updates an existing TPM group by its name. You cannot use the update endpoint to create new groups.
```

> Providing `member_tpm_ids` replaces the existing member list entirely. Providing
> `metadata` replaces the existing metadata map entirely.

Reviewer: ```suggestion
Vault entirely replaces the existing group member IDs with `member_tpm_ids` and
the existing metadata map with `metadata`. To avoid updating the group IDs or
metadata, either leave those fields unset or fully set them with the existing
information.
```

Author: This is a better formulation.  Except I'd prefer to remove "or fully set them with the existing information" because that's racy if someone else is making updates at the same time.  That ok with you?

Reviewer: sgtm 👍🏽

| Method | Path                             |
| :----- | :------------------------------- |
| `POST` | `/identity/tpmgroup/name/:name`  |

### Parameters

- `name` `(string: <required>)` – Name of the TPM group to update.

- `metadata` `(key-value-map: {})` – Updated metadata. Replaces existing
  metadata entirely.

- `member_tpm_ids` `(comma-separated string list: [])` – Updated list of
  member TPM IDs. Replaces the existing member list entirely. Every ID must
  reference an existing TPM record in the current namespace.

### Sample payload

```json
{
  "metadata": {
    "environment": "production",
    "version": "2"
  },
  "member_tpm_ids": [
    "sha256-a3f1c2d4e5b6...",
    "sha256-7d8e9f0a1b2c..."
  ]
}
```

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     http://127.0.0.1:8200/v1/identity/tpmgroup/name/datacenter-east-tpms

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/name/datacenter-east-tpms
```

```

### Sample response

```json
{
  "data": {

>     "id": "f47ac10b-58cc-4372-a567-0e02b2c3d479",

Reviewer: ```suggestion
    "id": "00000000-0000-0000-0000-000000000000",
```

    "name": "datacenter-east-tpms"
  }
}
```

## Delete TPM group by name

This endpoint deletes a TPM group by its name. Deleting a group that does not

> exist is idempotent and returns no error.

Reviewer: ```suggestion
Deletes a TPM group by its identifier. Deleting a group that does
not exist is idempotent and returns successfully as a no-op.
```

| Method   | Path                             |
| :------- | :------------------------------- |
| `DELETE` | `/identity/tpmgroup/name/:name`  |

### Parameters

- `name` `(string: <required>)` – Name of the TPM group.

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request DELETE \

>     http://127.0.0.1:8200/v1/identity/tpmgroup/name/datacenter-east-tpms

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/name/datacenter-east-tpms
```

```

## List TPM groups by name

> This endpoint returns a list of all TPM group names in the current namespace.

Reviewer: ```suggestion
Fetch a list of all TPM group names in the current namespace.
```

| Method | Path                                 |
| :----- | :----------------------------------- |
| `LIST` | `/identity/tpmgroup/name`            |
| `GET`  | `/identity/tpmgroup/name?list=true`  |

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request LIST \

>     http://127.0.0.1:8200/v1/identity/tpmgroup/name

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/identity/tpmgroup/name
```
